### ISSUE TITLE
Resolve social auth migration issue in production

### ISSUE THREAD

**Issue by raymondberg** [chicagopython/chipy.org#242]
We seem to be having the same problem in production as documented here: https://github.com/omab/python-social-auth/issues/682

This our output when migrating:

``
Operations to perform:
  Synchronize unmigrated apps: django_gravatar, django_ical, rest_framework, main, honeypot, django_extensions, admin_tools, humanize, about, interval, django_forms_bootstrap, thumbnail, ckeditor, nocaptcha_recaptcha, theming, messages, django_bleach, storages, tinymce, gunicorn, staticfiles, envelope, contact
  Apply all migrations: profiles, flatpages, admin, sessions, announcements, menu, contenttypes, sponsors, dashboard, auth, sites, social_django, subgroups, flatblocks, meetings
Synchronizing apps without migrations:
  Creating tables...
    Running deferred SQL...
  Installing custom SQL...
Running migrations:
  Rendering model states... DONE
  Applying meetings.0009_auto_20191231_1735... OK
  Applying meetings.0010_auto_20191231_1757... OK
  Applying profiles.0002_auto_20191231_1735... OK
  Applying social_django.0001_initial...Traceback (most recent call last):
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 62, in execute
    return self.cursor.execute(sql)
psycopg2.ProgrammingError: relation "social_auth_association" already exists

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "manage.py", line 9, in <module>
    execute_from_command_line(sys.argv)
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/commands/migrate.py", line 222, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 110, in migrate
    self.apply_migration(states[migration], migration, fake=fake, fake_initial=fake_initial)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/migrations/executor.py", line 148, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/migrations/migration.py", line 115, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/migrations/operations/models.py", line 59, in database_forwards
    schema_editor.create_model(model)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/base/schema.py", line 286, in create_model
    self.execute(sql, params or None)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/base/schema.py", line 111, in execute
    cursor.execute(sql, params)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/app/.local/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 62, in execute
    return self.cursor.execute(sql)
django.db.utils.ProgrammingError: relation "social_auth_association" already exists
```

**Comment by raymondberg** [comment]
For now, the workaround is to apply migrations from apps individually in production.

**Comment by elmq0022** [comment]
Any chance of just "faking" the migration migration that fails?  The fields are there so no need to create then.  

https://blog.dbrgn.ch/2014/10/30/migrating-django-social-auth/

**Comment by elmq0022** [comment]
I made some progress with this issue.  We won't be able to "fake" the migration.  The database and migrations are inconsistent.  

Results after faking the first migration: 

```shell
social_django
[X] 0001_initial (2 squashed migrations)
[ ] 0002_add_related_name (2 squashed migrations)
[ ] 0003_alter_email_max_length (2 squashed migrations)
[ ] 0004_auto_20160423_0400 (2 squashed migrations)
[ ] 0005_auto_20160727_2333 (1 squashed migrations)
[ ] 0006_partial
[ ] 0007_code_timestamp
[ ] 0008_partial_timestamp
``` 

```shell
Operations to perform:
  Apply all migrations: social_django
 Running migrations:
  Rendering model states... DONE
  Applying social_django.0002_add_related_name... OK
  Applying social_django.0003_alter_email_max_length...Traceback (most recent call last):
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
psycopg2.ProgrammingError: relation "social_auth_code" does not exist
```

The value for social_auth_code does not exist.

**Comment by elmq0022** [comment]
I have a preliminary solution :smiley:.  There seems to be just one table missing from the initial migration.  

### Step 1 - fake the initial migration
```
python manage.py migrate --fake social_django 0001_initial
```

### Step 2 - run the following SQL to create the missing table  
```
CREATE TABLE "social_auth_code" (
    "id" serial NOT NULL PRIMARY KEY, 
    "email" varchar(75) NOT NULL, 
    "code" varchar(32) NOT NULL, 
    "verified" boolean NOT NULL
);

ALTER TABLE "social_auth_code" ADD CONSTRAINT "social_auth_code_email_116e94140c869d48_uniq" UNIQUE ("email", "code");

COMMIT; 
```

### Step 3 - migrate as usual
```
python manage.py migrate
```
I'll do some additional checks, but I think we're back in business.  Hopefully, this won't mess up the existing data.

**Comment by elmq0022** [comment]
@raymondberg and @smandekar1, 

I was able to create a couple new oauth accounts just fine this morning.   Could one of you verify the steps above before we run it in production? 

Also, I have a new pull request to add a make psql shell, #243.  The command runs:

```shell
docker-compose exec db psql chipy chipy
``` 
You could use this psql shell to run the SQL on your dev machine.  This sql was generated from:  

```shell
python manage.py sqlmigrate social_django 0001_initial
```
and then trimmed back to the missing table.  

It's extremely similar to https://github.com/omab/python-social-auth/issues/682, but

**Comment by smandekar1** [comment]
getting this after step 1.    using the django-upgrade-2 branch
```
sameetmahisekar@MacBookPro chipy.org % make shell                                    
Opening shell in docker container
Use this shell to run python and django commands normally
app@be65cc55f81f:/site/proj$ python manage.py migrate --fake social_django 0001_initial
/home/app/.local/lib/python3.6/site-packages/interval/forms.py:5: RemovedInDjango19Warning: The django.forms.util module has been renamed. Use django.forms.utils instead.
  from django.forms.util import ValidationError

/home/app/.local/lib/python3.6/site-packages/interval/forms.py:28: RemovedInDjango19Warning: SortedDict is deprecated and will be removed in Django 1.9.
  ('X', 'microseconds')])

Operations to perform:
  Target specific migration: 0001_initial, from social_django
Running migrations:
  No migrations to apply.
Traceback (most recent call last):
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
psycopg2.ProgrammingError: relation "django_content_type" does not exist
LINE 1: ..."."app_label", "django_content_type"."model" FROM "django_co...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/app/.local/lib/python3.6/site-packages/django/contrib/contenttypes/models.py", line 67, in get_for_model
    ct = self.get(app_label=opts.app_label, model=opts.model_name)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/models/manager.py", line 127, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/models/query.py", line 328, in get
    num = len(clone)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/models/query.py", line 144, in __len__
    self._fetch_all()
  File "/home/app/.local/lib/python3.6/site-packages/django/db/models/query.py", line 965, in _fetch_all
    self._result_cache = list(self.iterator())
  File "/home/app/.local/lib/python3.6/site-packages/django/db/models/query.py", line 238, in iterator
    results = compiler.execute_sql()
  File "/home/app/.local/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 840, in execute_sql
    cursor.execute(sql, params)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/app/.local/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: relation "django_content_type" does not exist
LINE 1: ..."."app_label", "django_content_type"."model" FROM "django_co...
                                                             ^


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "manage.py", line 9, in <module>
    execute_from_command_line(sys.argv)
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/commands/migrate.py", line 226, in handle
    emit_post_migrate_signal(created_models, self.verbosity, self.interactive, connection.alias)
  File "/home/app/.local/lib/python3.6/site-packages/django/core/management/sql.py", line 280, in emit_post_migrate_signal
    using=db)
  File "/home/app/.local/lib/python3.6/site-packages/django/dispatch/dispatcher.py", line 189, in send
    response = receiver(signal=self, sender=sender, **named)
  File "/home/app/.local/lib/python3.6/site-packages/django/contrib/auth/management/__init__.py", line 82, in create_permissions
    ctype = ContentType.objects.db_manager(using).get_for_model(klass)
  File "/home/app/.local/lib/python3.6/site-packages/django/contrib/contenttypes/models.py", line 80, in get_for_model
    "Error creating new content types. Please make sure contenttypes "
RuntimeError: Error creating new content types. Please make sure contenttypes is migrated before trying to migrate apps individually.
app@be65cc55f81f:/site/proj$ 
```

**Comment by smandekar1** [comment]
I was getting this after step 1. running off the chipy master.  I tried migrating contenttypes alone and then step 1 worked.  
```
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/app/.local/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/app/.local/lib/python3.6/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: relation "django_content_type" does not exist
```
step 2 returned ```django.db.utils.ProgrammingError: relation "social_auth_partial" already exists```

**Comment by elmq0022** [comment]
@smandekar1 you will need to start with a clean database and back on tag 1.12.1 prior to Django social auth to python social auth migration.  Then mock the upgrade. 

 You could nuke the current volume, docker-compse volume rm chipy...    And docker-compse image rm chipy...  Then rebuild the environment.  

Alternatively build out a new environment typing the commands in the make file but adding the -p option.  Example docker-compse -p <name> up - - build -d.  -p let's you provide a name other than that of the parent directory.  

Rebuilding a docker-compse env does not delete the data from the volume by design. 

Step 2 is running SQL directly on postgresql and does not involve social auth partial.  So the error you got looks odd to me.

**Comment by elmq0022** [comment]
Maybe we can test it on your machine Thursday night before running it on the production site.

**Comment by smandekar1** [comment]
Sounds good, we can test it on this.  I did delete the database, but I didn't delete the image and start on 1.12.1
